web_src/js/features/repo-diff-selection.ts

```diff
@@ -0,0 +1,241 @@
+    return false;
+  }
+
+  const container = targetSpan.closest<HTMLElement>('.diff-file-box');
+  if (!container) return false;
+
+  // Check if the file is collapsed and expand it if needed
+  if (container.getAttribute('data-folded') === 'true') {
+    const foldBtn = container.querySelector<HTMLElement>('.fold-file');
+    if (foldBtn) {
+      // Expand the file using the setFileFolding utility
+      setFileFolding(container, foldBtn, false);
+      // Wait a bit for the expansion animation
+      await sleep(100);
+    }
+  }
+
+  if (!applyDiffLineSelection(container, range, {updateHash: false})) return false;
+  diffSelectionStart = {
+    anchor: targetId,
+    fragment: range.fragment,
+    side: range.startSide,
+    line: range.startLine,
+    container,
+  };
+
+  // Scroll to the first selected line (scroll to the tr element, not the span)
+  // The span is an inline element inside td, we need to scroll to the tr for better visibility
+  await sleep(10);
+  const targetTr = targetSpan.closest('tr');
+  if (targetTr) {
+    targetTr.scrollIntoView({behavior: 'smooth', block: 'center'});
+  }
+  return true;
+}
+
+function handleDiffLineNumberClick(cell: HTMLElement, e: MouseEvent) {
+  let span = cell.querySelector<HTMLSpanElement>('span[id^="diff-"]');
+  let info = parseDiffAnchor(span?.id ?? null);
+
+  // If clicked cell has no line number (e.g., clicking on the empty side of a deletion/addition),
+  // try to find the line number from the sibling cell on the same row
+  if (!info) {
+    const row = cell.closest('tr');
+    if (!row) return;
+    // Find the other line number cell in the same row
+    const siblingCell = cell.classList.contains('lines-num-old') ?
+      row.querySelector<HTMLElement>('td.lines-num-new') :
+      row.querySelector<HTMLElement>('td.lines-num-old');
+    if (siblingCell) {
+      span = siblingCell.querySelector<HTMLSpanElement>('span[id^="diff-"]');
+      info = parseDiffAnchor(span?.id ?? null);
+    }
+    if (!info) return;
+  }
+
+  const container = cell.closest<HTMLElement>('.diff-file-box');
+  if (!container) return;
+
+  e.preventDefault();
+
+  // Check if clicking on a single already-selected line without shift key - deselect it
+  if (!e.shiftKey) {
+    const clickedRow = cell.closest('tr');
+    if (clickedRow?.classList.contains('active')) {
+      // Check if this is a single-line selection by checking if it's the only selected line
+      const selectedRows = container.querySelectorAll('.code-diff tr.active');
+      if (selectedRows.length === 1) {
+        // This is a single selected line, deselect it
+        clickedRow.classList.remove('active');
+        diffSelectionStart = null;
+        // Remove hash from URL completely
+        if (window.history.pushState) {
+          window.history.pushState(null, null, window.location.pathname + window.location.search);
+        } else {
+          window.location.hash = '';
+        }
```

Comment:
Maybe you actually want to use [`window.history.replaceState`](https://developer.mozilla.org/en-US/docs/Web/API/History/replaceState)? I hate sites that push useless history entries because it pollutes the "back" stack, making the user hit the back button more than is necessary.